File: JINSEOP/프로그래머스_WEEK11/다트 게임.cpp
#include <vector>
#include <string>
using namespace std;

vector<int> score_list;
vector<char> bonus_list;
vector<pair<char, int>> option_list;

int str_to_int(string str) {
    if (str == "10") return 10;
    else return stoi(str);
}

void parse_dart(string dartResult) {
    string num_temp = "";

    for (auto c : dartResult) {
        if (c >= '0' && c <= '9')
            num_temp += c;
        if (c == 'S' || c == 'D' || c == 'T') {
            score_list.push_back(str_to_int(num_temp));  
            num_temp = "";
            bonus_list.push_back(c);
        }
        if (c == '*' || c == '#') option_list.push_back({c, bonus_list.size()-1});
    }
}

int calc_score() {
    int sum = 0;

    for (int i = 0; i < score_list.size(); i++) {
        switch (bonus_list[i])
        {
        case 'S':
            break;
        case 'D':
            score_list[i] = (score_list[i] * score_list[i]);
            break;
        case 'T':
            score_list[i] = (score_list[i] * score_list[i] * score_list[i]);
            break;
        }
    }

    for (int i = 0; i < option_list.size(); i++) {
        switch (option_list[i].first)
        {
        case '*':
            if (option_list[i].second == 0)
                score_list[option_list[i].second] *= 2;
            else {
                score_list[option_list[i].second - 1] *= 2;
                score_list[option_list[i].second] *= 2;
            }
            break;
        case '#':
            score_list[option_list[i].second] *= -1;
            break;
        }
    }

    for (int i : score_list) sum += i;

    return sum;
}

int solution(string dartResult) {
    int answer = 0;

    parse_dart(dartResult);
    answer = calc_score();
    return answer;
}

File: DONGUK/프로그래머스_WEEK_11/다트 게임.py
def solution(dartResult):
    answer = []
    
    i = 0
    prev_num = 0
    while i < len(dartResult):
        c = dartResult[i]
        
        if '0' <= dartResult[i+1] <= '9':
            c += dartResult[i+1]
            i += 1
        
        if '0' <= c[0] <= '9':
            tmp = ''
            i += 1
            while 'A' <= dartResult[i] <= 'Z' or dartResult[i] == '#' or  dartResult[i] == '*':
                tmp += dartResult[i]
                i += 1
                
                if i == len(dartResult):
                    break
            
            num = int(c)
            add = 0
            for j in range(len(tmp)):
                if tmp[j] == 'S':
                    add += (num)
                    
                elif tmp[j] == 'D':
                    add += (num ** 2)
                    
                elif tmp[j] == 'T':
                    add += (num ** 3)
                    
                elif tmp[j] == '*':
                    if i == 3:
                        add = add * 2
                    else:
                        answer[-1] = answer[-1] * 2
                        add = add * 2
                        
                elif tmp[j] == '#':
                    add = add * -1
                    
            answer.append(add)
    
    answer = sum(answer)
    return answer